### ISSUE TITLE
"stack level too deep" error when reading text on particular page

### ISSUE THREAD

**Issue by jokester** [yob/pdf-reader#94]
I encountered this with pdf-reader on linux.

It is page 775 in [this file](http://aclweb.org/anthology/C/C12/C12-1.pdf)

Code to reproduct it:

```
require "pdf-reader"
reader = PDF::Reader.new "C12-1.pdf"
reader.page(774).text # works
reader.page(775).text # raises the error
```

version: `pdf-reader 1.3.3` and `ruby 2.0.0p0`

**Comment by fro** [comment]
Same problem here:

    /Users/francois/.rbenv/versions/2.2.0-preview1/gemsets/playground-gems/gems/pdf-reader-1.3.3/lib/pdf/reader/object_cache.rb:71:in `include?': stack level too deep (SystemStackError)
    from /Users/francois/.rbenv/versions/2.2.0-preview1/gemsets/playground-gems/gems/pdf-reader-1.3.3/lib/pdf/reader/object_cache.rb:71:in `include?'
    from /Users/francois/.rbenv/versions/2.2.0-preview1/gemsets/playground-gems/gems/pdf-reader-1.3.3/lib/pdf/reader/object_hash.rb:79:in `[]'
    from /Users/francois/.rbenv/versions/2.2.0-preview1/gemsets/playground-gems/gems/pdf-reader-1.3.3/lib/pdf/reader/object_hash.rb:97:in `object'
    from /Users/francois/.rbenv/versions/2.2.0-preview1/gemsets/playground-gems/gems/pdf-reader-1.3.3/lib/pdf/reader/object_hash.rb:105:in `deref!'
    from /Users/francois/.rbenv/versions/2.2.0-preview1/gemsets/playground-gems/gems/pdf-reader-1.3.3/lib/pdf/reader/object_hash.rb:109:in `block (2 levels) in deref!'
    from /Users/francois/.rbenv/versions/2.2.0-preview1/gemsets/playground-gems/gems/pdf-reader-1.3.3/lib/pdf/reader/object_hash.rb:108:in `each'
    from /Users/francois/.rbenv/versions/2.2.0-preview1/gemsets/playground-gems/gems/pdf-reader-1.3.3/lib/pdf/reader/object_hash.rb:108:in `block in deref!'
    from /Users/francois/.rbenv/versions/2.2.0-preview1/gemsets/playground-gems/gems/pdf-reader-1.3.3/lib/pdf/reader/object_hash.rb:107:in `tap'
    ... 9295 levels...
    from pdf_customizer.rb:89:in `new'
    from pdf_customizer.rb:89:in `block in <main>'
    from pdf_customizer.rb:88:in `each'
    from pdf_customizer.rb:88:in `<main>'

**Comment by bcoles** [comment]
Same issue here.

* Ruby version 2.3.1
* pdf-reader version 1.4.1
* Ubuntu Linux

**Steps to reproduce:**

```bash
$ wget http://www.asx.com.au/asxpdf/20161128/pdf/43d893lymjk2vr.pdf
```

```ruby
#!/usr/bin/env ruby
require 'pdf-reader'

begin
  reader = PDF::Reader.new('43d893lymjk2vr.pdf')
rescue => e
  puts "Could not parse PDF: #{e}"
end

puts 'Read PDF successfully'

reader.pages.each do |page|
  page.text
end
```

```
$ ./bug.rb 
Read PDF successfully
/usr/local/rvm/gems/ruby-2.3.1/gems/pdf-reader-1.4.1/lib/pdf/reader/reference.rb:62:in `hash': stack level too deep (SystemStackError)
	from /usr/local/rvm/gems/ruby-2.3.1/gems/hashery-2.1.2/lib/hashery/lru_hash.rb:138:in `has_key?'
	from /usr/local/rvm/gems/ruby-2.3.1/gems/hashery-2.1.2/lib/hashery/lru_hash.rb:138:in `has_key?'
	from /usr/local/rvm/gems/ruby-2.3.1/gems/pdf-reader-1.4.1/lib/pdf/reader/object_cache.rb:71:in `include?'
	from /usr/local/rvm/gems/ruby-2.3.1/gems/pdf-reader-1.4.1/lib/pdf/reader/object_cache.rb:96:in `update_stats'
	from /usr/local/rvm/gems/ruby-2.3.1/gems/pdf-reader-1.4.1/lib/pdf/reader/object_cache.rb:28:in `[]'
	from /usr/local/rvm/gems/ruby-2.3.1/gems/pdf-reader-1.4.1/lib/pdf/reader/object_hash.rb:80:in `[]'
	from /usr/local/rvm/gems/ruby-2.3.1/gems/pdf-reader-1.4.1/lib/pdf/reader/object_hash.rb:97:in `object'
	from /usr/local/rvm/gems/ruby-2.3.1/gems/pdf-reader-1.4.1/lib/pdf/reader/object_hash.rb:105:in `deref!'
	 ... 10631 levels...
	from /usr/local/rvm/gems/ruby-2.3.1/gems/pdf-reader-1.4.1/lib/pdf/reader/page.rb:81:in `text'
	from ./bug.rb:17:in `block in <main>'
	from ./bug.rb:16:in `each'
	from ./bug.rb:16:in `<main>'

```

**Comment by yob** [comment]
Thanks for the detailed reproduction steps @bcoles. I identified the underlying issue in that ASX PDF, and built a [simple sample file and spec that triggers the bug](https://github.com/yob/pdf-reader/commit/f7a2aa39bff86eb085106cbdb24b375baaff2ee7).

I don't have a fix yet, but hopefully the hard work is behind me now.

**Comment by yob** [comment]
@jokester - now that I have a better idea of what's causing this, I was able to confirm that page 775 of the PDF you provided has the same "Recursive XObject" issue.

Presumably this discovery is too late to be helpful to you, but hopefully I can find a fix soon.

**Comment by jokester** [comment]
@yob thanks /o\ it's good to know that.

**Comment by yob** [comment]
I believe this has been resolved in #198 - please let me know if I'm wrong!

**Comment by bcoles** [comment]
Thanks @yob 

The issue appears to be resolved in the example files provided by @jokester @driki and myself.

Tested on Ruby 2.3.1
